include/deal.II/base/hdf5.h

```diff
@@ -0,0 +1,1073 @@
+     */
+    unsigned int rank;
+
+    /**
+     * The vector `dimensions` is a one-dimensional array of size rank
+     * specifying the size of each dimension of the dataset.
+     */
+    std::vector<hsize_t> dimensions;
+
+    /**
+     * HDF5 dataspace identifier.
+     */
+    std::shared_ptr<hid_t> dataspace;
+
+    /**
+     * Total number of elements in the dataset.
+     */
+    unsigned int size;
+
+    /**
+     * If query_io_mode is set to true, then after every read/write operation
+     * the HDF5 deal.ii interface calls the routines
+     * [H5Pget_mpio_actual_io_mode()](https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-GetMpioActualIoMode)
+     * and
+     * [H5Pget_mpio_no_collective_cause()](https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-GetMpioNoCollectiveCause).
+     * The results are stored in io_mode, local_no_collective_cause and
+     * global_no_collective_cause.
+     */
+    bool query_io_mode;
+
+    /**
+     * I/O mode that was performed on the last parallel I/O call.
+     */
+    H5D_mpio_actual_io_mode_t io_mode;
+
+    /**
+     * Local causes that broke collective I/O on the
+     * last parallel I/O call. See <a
+     * href="https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-GetMpioNoCollectiveCause">H5Pget_mpio_no_collective_cause</a>.
+     */
+    uint32_t local_no_collective_cause;
+
+    /**
+     * Global causes that broke collective I/O on the
+     * last parallel I/O call. See <a
+     * href="https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-GetMpioNoCollectiveCause">H5Pget_mpio_no_collective_cause</a>.
+     */
+    uint32_t global_no_collective_cause;
+  };
+
+  /**
+   * This class implements an HDF5 Group
+   *
+   * @author Daniel Garcia-Sanchez, 2018
+   */
+  class Group : public HDF5Object
+  {
+  protected:
+    /**
+     * Group access mode
+     */
+    enum class GroupAccessMode
+    {
+      /**
+       * Opens an existing group
+       */
+      open,
+      /**
+       * Creates a new group
+       */
+      create
+    };
+    /**
+     * This constructor creates or opens a group depending on the value of
+     * @p mode. The group will be placed insided the group @p parent_group. The
+     * parameter @p mpi defines if the the I/O operations are serial or
+     * parallel. This is an internal constructor, the functions open_group() and
+     * create_group() should be used to open or create a group.
+     */
+    Group(const std::string &   name,
+          const Group &         parent_group,
+          const bool            mpi,
+          const GroupAccessMode mode);
+
+    /**
+     * Internal constructor used by File. The constructor sets the protected
+     * const members of HDF5Group: @p name and @p mpi. It does not create or
+     * open a Group.
+     */
+    Group(const std::string &name, const bool mpi);
+
+  public:
+    /**
+     * Opens a group.
+     */
+    Group
+    open_group(const std::string &name);
```

Comment:
Since you're already in a `Group` class, does this mean that groups can be nested? In other words, that a group can contain a group? If so, maybe augment the documentation to say "Opens a sub-group of the current group."